README.rst
Django-int-money
================

Rewrite/Fork of Django-money, with some key differences:

- simplified code base
- store money amounts as integers
- store currency information in a table rather than a field

Rationale
---------
Storing money values as integers in the database has several advantages:

- takes up less space in the database
- avoids any issues with rounding or conversion errors compared to floats
  (don't use floats for money values, **ever**!)
- allows using the datbase to perform SUMs, AVGs, etc more efficiently compared to NUMERIC / DECIMAL types
- easier to import or export data to other systems

Storing the currency information at the table level has several advantages:

- takes up less space in the database: not needing a char(3) field for every amount to be stored
  can have a significant impact. Consider a table with 5 money fields containing a million entries,
  that's about a 20 Mb difference.
- from a data modeling perspective, it's best to completely separate different currency types.
  There's no valid reason to ever do calculations on different currencies.

On the other hand, this requires a little more planning ahead when designing
the data model, since adding a currency to the application requires adding a
table.

To alleviate this problem, table inheritence is used to avoid needless repetition.


Requirements
------------
Django versions: 1.7, 1.8, 1.9

Python versions: 3.3, 3.4, 3.5

PyPy versions: PyPy3 2.4

Yeah, that's right: there's no official Python 2 support.
We are now living in the ``__future__`` ;-)

Via ``py-moneyed``, ``django-int-money`` gets:

-  Support for proper Money value handling (using the standard Money
   design pattern)
-  A currency class and definitions for all currencies in circulation
-  Formatting of most currencies with correct currency sign


Installation
------------

Django-int-money currently needs ``py-moneyed`` v0.4 (or later) to work.

You can obtain the source code for ``django-int-money`` from here:

::

    https://github.com/ianare/django-int-money

And the source for ``py-moneyed`` from here:

::

    https://github.com/limist/py-moneyed

Using `pip`:

    pip install py-moneyed django-int-money

Model Usage
-----------

Use as normal model fields

.. code:: python

        from dj_intmoney.models.fields import MoneyIntegerField
        from django.db import models

        class BankAccountUSD(models.Model):

            balance = MoneyIntegerField()

            @staticmethod
            def get_currency():
                return "USD"


        class BankAccountEUR(models.Model):

            balance = MoneyIntegerField()

            @staticmethod
            def get_currency():
                return "EUR"


Support for Django migrations built in.


Adding a new Currency
---------------------

Currencies are listed on moneyed, and this modules use this to provide a
choice list on the admin, also for validation.

To add a new currency available on all the project, you can simple add
this two lines on your ``settings.py`` file

.. code:: python

        import moneyed
        from moneyed.localization import _FORMATTER
        from decimal import ROUND_HALF_EVEN

        BOB = moneyed.add_currency(
            code='BOB',
            numeric='068',
            name='Peso boliviano',
            countries=('BOLIVIA', )
        )

        # Currency Formatter will output 2.000,00 Bs.
        _FORMATTER.add_sign_definition(
            'default',
            BOB,
            prefix=u'Bs. '
        )

        _FORMATTER.add_formatting_definition(
            'es_BO',
            group_size=3, group_separator=".", decimal_point=",",
            positive_sign="",  trailing_positive_sign="",
            negative_sign="-", trailing_negative_sign="",
            rounding_method=ROUND_HALF_EVEN)


Important note on model managers
--------------------------------

Django-money leaves you to use any custom model managers you like for
your models, but it needs to wrap some of the methods to allow searching
for models with money values.

This is done automatically for the "objects" attribute in any model that
uses MoneyIntegerField. However, if you assign managers to some other
attribute, you have to wrap your manager manually, like so:

.. code:: python

        from djmoney.models.managers import money_manager
        class BankAccount(models.Model):

            balance = MoneyIntegerField(max_digits=10, decimal_places=2, default_currency='USD')

            accounts = money_manager(MyCustomManager())

Also, the money\_manager wrapper only wraps the standard QuerySet
methods. If you define custom QuerySet methods, that do not end up using
any of the standard ones (like "get", "filter" and so on), then you also
need to manually decorate those custom methods, like so:

.. code:: python

        from djmoney.models.managers import understand_money

        class MyCustomQuerySet(QuerySet):

           @understand_money
           def my_custom_method(*args,**kwargs):
               # Awesome stuff

Format localization
-------------------

The formatting is turned on if you have set ``USE_L10N = True`` in the
your settings file.

If formatting is disabled in the configuration, then in the templates
will be used default formatting.

In the templates you can use a special tag to format the money.

In the file ``settings.py`` add to ``INSTALLED_APPS`` entry from the
library ``dj_intmoney``:

.. code:: python

        INSTALLED_APPS += ( 'dj_intmoney', )

In the template, add:

::

        {% load dj_intmoney %}
        ...
        {% money_localize money %}

and that is all.

Instructions to the tag ``money_localize``:

::

            {% money_localize <money_object> [ on(default) | off ] [as var_name] %}
            {% money_localize <amount> <currency> [ on(default) | off ] [as var_name] %}

Examples:

The same effect:

::

            {% money_localize money_object %}
            {% money_localize money_object on %}

Assignment to a variable:

::

            {% money_localize money_object on as NEW_MONEY_OBJECT %}

Formatting the number with currency:

::

            {% money_localize '4.5' 'USD' %}

::

    Return::

        MoneyPatched object

Testing
-------

Install the required packages:

::

    git clone https://github.com/ianare/django-int-money

    cd ./django-money/

    pip install -e .[tests] # installation with required packages for testing

Recommended way to run the tests:

.. code:: bash

    tox

Testing the application in the current environment python:

.. code:: bash

    make test

Working with Exchange Rates
---------------------------

To work with exchange rates, check out this repo that builds off of
django-money: https://github.com/evonove/django-money-rates



CHANGES.rst
v0.0.1
------
- Initial release (beta)


dj_intmoney/models/fields.py
# -*- coding: utf-8 -*-
from __future__ import division

import inspect
from decimal import ROUND_DOWN, Decimal

from django.conf import settings
from django.core.exceptions import ValidationError
from django.db import models
from django.db.models import F
from django.db.models.signals import class_prepared
from django.utils import translation
from django.utils.encoding import smart_text

from moneyed import Currency, Money
from moneyed.localization import _FORMATTER, format_money

from .. import forms

from .._compat import (
    BaseExpression,
    Expression,
    deconstructible,
    split_expression,
    string_types,
)


# If django-money-rates is installed we can automatically
# perform operations with different currencies
if 'djmoney_rates' in settings.INSTALLED_APPS:
    try:
        from djmoney_rates.utils import convert_money
        AUTO_CONVERT_MONEY = True
    except ImportError:
        # NOTE. djmoney_rates doesn't support Django 1.9+
        AUTO_CONVERT_MONEY = False
else:
    AUTO_CONVERT_MONEY = False


__all__ = ('MoneyIntegerField', 'NotSupportedLookup')

SUPPORTED_LOOKUPS = ('exact', 'isnull', 'lt', 'gt', 'lte', 'gte')


class NotSupportedLookup(Exception):

    def __init__(self, lookup):
        self.lookup = lookup

    def __str__(self):
        return 'Lookup \'%s\' is not supported for MoneyIntegerField' % self.lookup


@deconstructible
class MoneyPatched(Money):

    # Set to True or False has a higher priority
    # than USE_L10N == True in the django settings file.
    # The variable "self.use_l10n" has three states:
    use_l10n = None

    def __float__(self):
        return float(self.amount)

    def _convert_to_local_currency(self, other):
        """
        Converts other Money instances to the local currency
        """
        if AUTO_CONVERT_MONEY:
            return convert_money(other.amount, other.currency, self.currency)
        else:
            return other

    @classmethod
    def _patch_to_current_class(cls, money):
        """
        Converts object of type MoneyPatched on the object of type Money.
        """
        return cls(money.amount, money.currency)

    def __pos__(self):
        return MoneyPatched._patch_to_current_class(
            super(MoneyPatched, self).__pos__())

    def __neg__(self):
        return MoneyPatched._patch_to_current_class(
            super(MoneyPatched, self).__neg__())

    def __add__(self, other):
        other = self._convert_to_local_currency(other)
        return MoneyPatched._patch_to_current_class(
            super(MoneyPatched, self).__add__(other))

    def __sub__(self, other):
        other = self._convert_to_local_currency(other)
        return MoneyPatched._patch_to_current_class(
            super(MoneyPatched, self).__sub__(other))

    def __mul__(self, other):

        return MoneyPatched._patch_to_current_class(
            super(MoneyPatched, self).__mul__(other))

    def __eq__(self, other):
        if hasattr(other, 'currency'):
            if self.currency == other.currency:
                return self.amount == other.amount
            raise TypeError('Cannot add or subtract two Money ' +
                            'instances with different currencies.')
        return False

    def __truediv__(self, other):
        if isinstance(other, Money):
            return super(MoneyPatched, self).__truediv__(other)
        else:
            return self._patch_to_current_class(
                super(MoneyPatched, self).__truediv__(other))

    def __rmod__(self, other):
        return MoneyPatched._patch_to_current_class(
            super(MoneyPatched, self).__rmod__(other))

    def __get_current_locale(self):
        # get_language can return None starting on django 1.8
        language = translation.get_language() or settings.LANGUAGE_CODE
        locale = translation.to_locale(language)

        if _FORMATTER.get_formatting_definition(locale):
            return locale

        if _FORMATTER.get_formatting_definition('%s_%s' % (locale, locale)):
            return '%s_%s' % (locale, locale)

        return ''

    def __use_l10n(self):
        """
        Return boolean.
        """
        if self.use_l10n is None:
            return settings.USE_L10N
        return self.use_l10n

    def __str__(self):
        if self.__use_l10n():
            locale = self.__get_current_locale()
            if locale:
                return format_money(self, locale=locale)

        return format_money(self)

    def __repr__(self):
        return '%s %s' % (self.amount.to_integral_value(ROUND_DOWN), self.currency)


def get_value(obj, expr):
    """
    Extracts value from object or expression.
    """
    if isinstance(expr, F):
        expr = getattr(obj, expr.name)
    elif hasattr(expr, 'value'):
        expr = expr.value
    return expr


def validate_money_expression(obj, expr):
    """
    Money supports different types of expressions, but you can't do following:
      - Add or subtract money with not-money
      - Any exponentiation
      - Any operations with money in different currencies
      - Multiplication, division, modulo with money instances on both sides of expression
    """
    lhs, rhs = split_expression(expr)
    connector = expr.connector
    lhs = get_value(obj, lhs)
    rhs = get_value(obj, rhs)

    if (not isinstance(rhs, Money) and connector in ('+', '-')) or connector == '^':
        raise ValidationError('Invalid F expression for MoneyIntegerField.', code='invalid')
    if isinstance(lhs, Money) and isinstance(rhs, Money):
        if connector in ('*', '/', '^', '%%'):
            raise ValidationError('Invalid F expression for MoneyIntegerField.', code='invalid')
        if lhs.currency != rhs.currency:
            raise ValidationError('You cannot use F() with different currencies.', code='invalid')


def validate_money_value(value):
    """
    Valid value for money are:
      - Single numeric value
      - Money instances
      - Pairs of numeric value and currency. Currency can't be None.
    """
    if isinstance(value, (list, tuple)) and (len(value) != 2 or value[1] is None):
        raise ValidationError(
            'Invalid value for MoneyIntegerField: %(value)s.',
            code='invalid',
            params={'value': value},
        )


def get_currency(value):
    """
    Extracts currency from value.
    """
    if isinstance(value, Money):
        return smart_text(value.currency)
    elif isinstance(value, (list, tuple)):
        return value[1]


class MoneyFieldProxy(object):

    def __init__(self, field):
        self.field = field

    def _money_from_obj(self, obj):
        amount = obj.__dict__[self.field.name]
        currency = self.field.model.get_currency()
        if amount is None:
            return None
        return MoneyPatched(amount=amount, currency=currency)

    def __get__(self, obj, type=None):
        if obj is None:
            raise AttributeError('Can only be accessed via an instance.')
        if isinstance(obj.__dict__[self.field.name], BaseExpression):
            return obj.__dict__[self.field.name]
        if not isinstance(obj.__dict__[self.field.name], Money):
            obj.__dict__[self.field.name] = self._money_from_obj(obj)
        return obj.__dict__[self.field.name]

    def __set__(self, obj, value):  # noqa
        # if isinstance(value, BaseExpression):
        #     validate_money_expression(obj, value)
        #     prepare_expression(value)
        # else:
        #     validate_money_value(value)
        #     currency = get_currency(value)
        #     if currency:
        #         self.set_currency(obj, currency)
        value = self.field.to_python(value)
        obj.__dict__[self.field.name] = value


class MoneyIntegerField(models.IntegerField):
    description = 'A field which stores both the currency and amount of money.'

    def __init__(self, verbose_name=None, name=None,
                 default=None, **kwargs):
        nullable = kwargs.get('null', False)
        if default is None and not nullable:
            # Backwards compatible fix for non-nullable fields
            default = 0

        default = Money(default, )

        if not (nullable and default is None) and not isinstance(default, Money):
            raise Exception(
                'default value must be an instance of Money, is: %s' % str(default))

        super(MoneyIntegerField, self).__init__(verbose_name, name, default=default, **kwargs)

    def to_python(self, value):
        if isinstance(value, Expression):
            return value
        if isinstance(value, Money):
            value = value.amount
        elif isinstance(value, float):
            value = Decimal.from_float(value)
        elif isinstance(value, str):
            value = Decimal(value)
        elif isinstance(value, int):
            value = Decimal(value) / 100
        return value
        #return super(MoneyIntegerField, self).to_python(value)

    def contribute_to_class(self, cls, name):
        if not cls._meta.abstract:
            cls._meta.has_money_field = True
        super(MoneyIntegerField, self).contribute_to_class(cls, name)
        setattr(cls, self.name, MoneyFieldProxy(self))

    def get_db_prep_save(self, value, connection):
        if isinstance(value, Expression):
            return value
        if isinstance(value, Money):
            value = int(value.amount * 100)
        elif isinstance(value, int):
            value = int(Decimal(value) * 100)
        return super(MoneyIntegerField, self).get_db_prep_save(value, connection)

    def get_db_prep_lookup(self, lookup_type, value, connection, prepared=False):
        if lookup_type not in SUPPORTED_LOOKUPS:
            raise NotSupportedLookup(lookup_type)
        value = self.get_db_prep_save(value, connection)
        return super(MoneyIntegerField, self).get_db_prep_lookup(lookup_type, value, connection, prepared)

    def get_default(self):
        if isinstance(self.default, Money):
            frm = inspect.stack()[1]
            mod = inspect.getmodule(frm[0])
            return self.default
        else:
            return super(MoneyIntegerField, self).get_default()

    def formfield(self, **kwargs):
        defaults = {'form_class': forms.MoneyField}
        defaults.update(kwargs)
        return super(MoneyIntegerField, self).formfield(**defaults)

    def value_to_string(self, obj):
        value = int(self._get_val_from_obj(obj).amount * 100)
        return self.get_prep_value(value)

    # Django 1.7 migration support
    def deconstruct(self):
        name, path, args, kwargs = super(MoneyIntegerField, self).deconstruct()

        if self.default is not None:
            kwargs['default'] = self.default.amount
        return name, path, args, kwargs


def patch_managers(sender, **kwargs):
    """
    Patches models managers
    """
    from .managers import money_manager

    if hasattr(sender._meta, 'has_money_field'):
        sender.copy_managers([
            (_id, name, money_manager(manager))
            for _id, name, manager in sender._meta.concrete_managers
        ])


class_prepared.connect(patch_managers)
